wl-sim/main/main.cpp

```diff
@@ -0,0 +1,191 @@
+int feistel_test();
+
+int main(int argc, char **argv)
+{
+    srand(time(0));
+    esp_err_t result;
+
+    // if single argument 'test', run the mapping correctness test
+    if (argc == 2 && strcmp(argv[1], "test") == 0) {
+        feistel_test();
+        return 0;
+    }
+
+    // otherwise require all args for a simulation run
+    // e.g. wl-sim f z z 10 0
+    // for Feistel enabled, zipf address access and zipf block size with maximum of 10 and 0 per mille chance for restart
+    if (argc != 6) {
+        printf("Need simulation params as arguments:\n\
+\tMAPPING_ALG: f for Feistel, b for base mapping alg\n\
+\tADDRESS_FUNC: z for zipf, c for const\n\
+\tBLOCKS_SIZE_FUNC: z for zipf, c for const\n\
+\tBLOCK_SIZE: N for max erase block size\n\
+\tRESTART_PROB: P for restart probability after every erase [per mille]\n");
+        return ESP_FAIL;
+    }
+
+    // argument parsing
+
+    bool enable_feistel = false;
+    if (strcmp(argv[1], "f") == 0) {
+        enable_feistel = true;
+    } else if (strcmp(argv[1], "b") == 0) {
+        // enable_feistel already false
+    } else {
+        fprintf(stderr, "First argument '%s' invalid, defaulting to base mapping...\n", argv[1]);
+        // enable_feistel already false
+    }
+
+    address_function_t addr_func = &constant;
+    if (strcmp(argv[2], "z") == 0) {
+        addr_func = &zipf;
+    } else if (strcmp(argv[2], "c") == 0) {
+        // addr_func already constant
+    } else {
+        fprintf(stderr, "Second argument '%s' invalid, defaulting to constant address...\n", argv[2]);
+        // addr_func already constant
+    }
+
+    block_size_function_t block_func = &block_constant;
+    if (strcmp(argv[3], "z") == 0) {
+        block_func = &block_zipf;
+    } else if (strcmp(argv[3], "c") == 0) {
+        // block_func already constant
+    } else {
+        fprintf(stderr, "Third argument '%s' invalid, defaulting to constant erase block size...\n", argv[3]);
+        // block_func already constant
+    }
+
+    // last two arguments are numbers
+    char *end = NULL;
+    int erase_block_size = strtol(argv[4], &end, 10);
+    if (*end != '\0') {
+        fprintf(stderr, "Invalid erase block size '%s'!\n", argv[4]);
+        return ESP_FAIL;
+    } else if (erase_block_size <= 0) {
+        fprintf(stderr, "Invalid erase block size %i, must be > 0\n", erase_block_size);
+        return ESP_FAIL;
+    }
+
+    int restart_prob = strtol(argv[5], &end, 10);
+    if (*end != '\0') {
+        fprintf(stderr, "Invalid restart probability %s'!\n", argv[5]);
+        return ESP_FAIL;
+    } else if (restart_prob < 0) {
+        fprintf(stderr, "Invalid restart probability %i, must be > 0 or 0 to turn off random restarting\n", restart_prob);
+        return ESP_FAIL;
+    }
+
+    // if Feistel enabled from args, init keys and variables
+    if (enable_feistel) {
+        init_feistel(false);
+    }
+
+    // runs until any sector reaches erase lifetime, see erase_sector()
+    for (;;) {
```

Comment:
Nitpicking:
this block could be redefined to something like

```
for (;; erase_range(addr_func(FLASH_SIZE), ERASE_SIZE * block_func(erase_block_size)) == ESP_OK) {

        // if nonzero restart probability from arguments
        if (restart_prob != 0) {
            // generate number P in per mille to compare with given restart prob
            int P = rand() % 1000;
            if (P < restart_prob) {
                // and simulate a restart, loosing current value of access_count
                access_count = 0;
                restarted++;
            }
        }
    }
``` 

to improve the code structure

Reply:
yeah, another thanks as that `for(;;)` approach was convoluted